File: service/system/dept.py
import uuid

from database.postgre_client import get_session
from database.system.dept import DeptRepository
from exception.bad_except import bad_except


class DeptService:
    """
    部门业务层。

    负责编排持久层调用与业务规则：父节点存在性校验、防环校验
    （新父不能是自身或后代）、删除守卫（无子部门且无关联用户）。
    列表为扁平结构（树由前端组装），每个方法在共享会话中开启事务并提交。
    """

    async def create(
        self,
        name: str,
        parent_id: uuid.UUID | None = None,
        sort_order: int = 0,
        leader: str | None = None,
        status: str = "active",
    ) -> dict:
        """创建部门并返回其数据；提供 parent_id 时校验父部门存在。"""
        async with get_session() as session:
            repo = DeptRepository(session)
            if parent_id is not None:
                parent = await repo.get(parent_id)
                if parent is None:
                    bad_except(f"父部门不存在: {parent_id}")
            dept = await repo.create(
                name=name,
                parent_id=parent_id,
                sort_order=sort_order,
                leader=leader,
                status=status,
            )
            await session.commit()
            return dept.to_dict()

    async def list(
        self,
        keyword: str | None = None,
        status: str | None = None,
    ) -> list[dict]:
        """扁平列表（sort_order 升序），供前端组树。"""
        async with get_session() as session:
            repo = DeptRepository(session)
            depts = await repo.list(keyword=keyword, status=status)
            return [d.to_dict() for d in depts]

    async def get(self, dept_id: uuid.UUID) -> dict:
        """按 id 获取部门，不存在时抛出业务异常。"""
        async with get_session() as session:
            repo = DeptRepository(session)
            dept = await repo.get(dept_id)
            if dept is None:
                bad_except(f"部门不存在: {dept_id}")
            return dept.to_dict()

    async def update(
        self,
        dept_id: uuid.UUID,
        name: str | None = None,
        sort_order: int | None = None,
        leader: str | None = None,
        status: str | None = None,
        parent_id: uuid.UUID | None = None,
        parent_id_set: bool = False,
    ) -> dict:
        """
        更新部门，不存在时抛出业务异常。

        变更 parent_id 时校验：父部门存在、不是自身、也不是自身的后代
        （防环，用内存 BFS 后代集合判断）。
        """
        async with get_session() as session:
            repo = DeptRepository(session)
            dept = await repo.get(dept_id)
            if dept is None:
                bad_except(f"部门不存在: {dept_id}")
            if parent_id_set and parent_id is not None:
                if parent_id == dept_id:
                    bad_except("父部门不能是自身")
                parent = await repo.get(parent_id)
                if parent is None:
                    bad_except(f"父部门不存在: {parent_id}")
                descendants = await repo.list_descendant_ids(dept_id)
                if parent_id in descendants:
                    bad_except("父部门不能是自身的下级部门")
            dept = await repo.update(
                dept_id,
                name=name,
                sort_order=sort_order,
                leader=leader,
                status=status,
                parent_id=parent_id,
                parent_id_set=parent_id_set,
            )
            await session.commit()
            return dept.to_dict()

    async def delete(self, dept_id: uuid.UUID) -> None:
        """带守卫的物理删除：存在子部门或关联用户时拒绝删除。"""
        async with get_session() as session:
            repo = DeptRepository(session)
            dept = await repo.get(dept_id)
            if dept is None:
                bad_except(f"部门不存在: {dept_id}")
            if await repo.has_children(dept_id):
                bad_except("部门下存在子部门，无法删除")
            if await repo.has_referencing_user(dept_id):
                bad_except("部门下存在用户，无法删除")
            await repo.delete(dept_id)
            await session.commit()

File: multi-agent-ui/src/api/system/dept.ts
import request, { type ApiResult } from "@/utils/request";
import { DEPT_URL } from "./index";

/** 部门实体（对应后端 sys_dept.to_dict） */
export interface Dept {
  id: string;
  /** 父部门 id，为空表示顶级部门 */
  parent_id: string | null;
  name: string;
  sort_order: number;
  leader: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

/** 部门树节点（在扁平 Dept 上追加 children，用于树形表格/树选择） */
export interface DeptTreeNode extends Dept {
  children: DeptTreeNode[];
}

/** 将扁平的 parent_id 部门列表组装成树，按 sort_order 升序排列同级节点 */
export function buildDeptTree(list: Dept[]): DeptTreeNode[] {
  const map = new Map<string, DeptTreeNode>();
  list.forEach((d) => map.set(d.id, { ...d, children: [] }));
  const roots: DeptTreeNode[] = [];
  map.forEach((node) => {
    const parent = node.parent_id ? map.get(node.parent_id) : null;
    if (parent) parent.children.push(node);
    else roots.push(node);
  });
  const sortNodes = (nodes: DeptTreeNode[]) => {
    nodes.sort((a, b) => a.sort_order - b.sort_order);
    nodes.forEach((n) => sortNodes(n.children));
  };
  sortNodes(roots);
  return roots;
}

/** 收集某个部门自身及其全部后代的 id（编辑部门时排除自引用与成环） */
export function collectDeptSubtreeIds(list: Dept[], rootId: string): Set<string> {
  const childrenMap = new Map<string | null, Dept[]>();
  list.forEach((d) => {
    const key = d.parent_id;
    const bucket = childrenMap.get(key);
    if (bucket) bucket.push(d);
    else childrenMap.set(key, [d]);
  });
  const ids = new Set<string>([rootId]);
  const stack = [rootId];
  while (stack.length) {
    const cur = stack.pop() as string;
    (childrenMap.get(cur) ?? []).forEach((child) => {
      if (!ids.has(child.id)) {
        ids.add(child.id);
        stack.push(child.id);
      }
    });
  }
  return ids;
}

/** 创建部门请求体（parent_id 为空表示顶级部门） */
export interface DeptCreatePayload {
  name: string;
  parent_id?: string | null;
  sort_order?: number;
  leader?: string | null;
  status?: string;
}

/** 更新部门请求体（仅提供的字段会被更新，parent_id 可显式置空升为顶级） */
export interface DeptUpdatePayload {
  name?: string;
  sort_order?: number;
  leader?: string | null;
  status?: string;
  parent_id?: string | null;
}

/** 扁平列出部门参数（不分页，树由前端组装） */
export interface DeptListParams {
  /** 按部门名称模糊搜索 */
  keyword?: string;
  /** 按状态精确过滤 */
  status?: string;
}

/** 部门管理 API：统一通过 deptApi.xx() 调用 */
export const deptApi = {
  /** 创建部门（父部门须存在） */
  create(payload: DeptCreatePayload) {
    return request.post<Dept, ApiResult<Dept>>(DEPT_URL.root, payload);
  },

  /** 扁平列出全部部门（sort_order 升序），树由前端 buildDeptTree 组装 */
  list(params: DeptListParams = {}) {
    return request.get<Dept[], ApiResult<Dept[]>>(DEPT_URL.root, { params });
  },

  /** 按 id 获取部门 */
  get(deptId: string) {
    return request.get<Dept, ApiResult<Dept>>(DEPT_URL.byId(deptId));
  },

  /** 更新部门；变更 parent_id 时后端校验存在性与防环（非自身/后代） */
  update(deptId: string, payload: DeptUpdatePayload) {
    return request.put<Dept, ApiResult<Dept>>(DEPT_URL.byId(deptId), payload);
  },

  /** 带守卫的物理删除：存在子部门或关联用户时拒绝删除 */
  remove(deptId: string) {
    return request.delete<null, ApiResult<null>>(DEPT_URL.byId(deptId));
  },
};
